cleanup of non-blocking code

dstaesse opened the pull request:
blocking property of flows is now restricted to flow allocation
by removing it from the application registration, apps can now
have both blocking and non-blocking i/o from flows. I only left
the blocking property in the FlowSpecification object, but it
should not be there either. But better to wait for filedescriptors
for flows to fix it completely. This is already a step towards that.

librina   : removed blocking property from registering applications
            and associated structures, netlink messages and APIs

linux     : removed blocking property in shims, kipcm and netlink

rina-tools: adapted tools to new API

rinad     : removed blocking=true from some register_to_dif calls
            flow allocator now accepts blocking in the API

MAINTAINERS:
F: librina/include/librina/common.h
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/include/librina/ipc-api.h
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/include/librina/ipc-manager.h
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/include/librina/ipc-process.h
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/src/common.cc
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/src/ipc-api.cc
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/src/ipc-manager.cc
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/src/ipc-process.cc
        Eduard Grasa <eduard.grasa (at) i2cat.net>
F: librina/src/netlink-messages.cc
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: librina/src/netlink-messages.h
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: librina/src/netlink-parsers.cc
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: librina/src/netlink-parsers.h
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: librina/src/syscalls-wrappers.cc
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: linux/net/rina/common.h
        Vincenzo Maffione <v.maffione (at) nextworks.it>
F: linux/net/rina/default.c
        Vincenzo Maffione <v.maffione (at) nextworks.it>
F: linux/net/rina/ipcps/shim-eth-vlan.c
        Sander Vrijders <sander.vrijders (at) intec.ugent.be>
F: linux/net/rina/ipcps/shim-hv.c
        Vincenzo Maffione <v.maffione (at) nextworks.it>
F: linux/net/rina/ipcps/shim-tcp-udp.c
        Sander Vrijders <sander.vrijders (at) intec.ugent.be>
F: linux/net/rina/kfa.c
        Miquel Tarzan <miquel.tarzan (at) i2cat.net>
F: linux/net/rina/kfa.h
        Miquel Tarzan <miquel.tarzan (at) i2cat.net>
F: linux/net/rina/kipcm.c
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: linux/net/rina/kipcm.h
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: linux/net/rina/personality.c
        Vincenzo Maffione <v.maffione (at) nextworks.it>
F: linux/net/rina/personality.h
        Vincenzo Maffione <v.maffione (at) nextworks.it>
F: linux/net/rina/rnl-utils.c
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: linux/net/rina/rnl-utils.h
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: linux/net/rina/syscalls.c
        Leonardo Bergesio <leonardo.bergesio (at) i2cat.net>
F: rinad/src/ipcm/flow-alloc-handlers.cc
        Marc Sune <marc.sune (at) bisdn.de>
F: rinad/src/ipcm/ipcp.cc
        Marc Sune <marc.sune (at) bisdn.de>
F: rinad/src/ipcm/ipcp.h
        Marc Sune <marc.sune (at) bisdn.de>

sandervrijders commented on linux/net/rina/ipcps/shim-eth-vlan.c:
```diff
@@ -51,6 +51,8 @@
 #define DEFAULT_QDISC_MAX_SIZE 50
 #define DEFAULT_QDISC_ENABLE_SIZE 10
 
+#define BLOCKING_FLOW 1
```
Can you elaborate why this is needed? Is this needed in every shim? Then probably it should be defined in a common place (common.h)?

edugrasa replied:
Blocking or non blocking is a property of flows consumed by apps (via the
read/write syscalls implented in the KFA), not of flows used by IPCPs.
El 17/08/2015 10:38, "Sander Vrijders" <notifications@github.com> escribió:

> In linux/net/rina/ipcps/shim-eth-vlan.c
> <https://github.com/IRATI/stack/pull/713#discussion_r37167575>:
>
> > @@ -51,6 +51,8 @@
> >  #define DEFAULT_QDISC_MAX_SIZE 50
> >  #define DEFAULT_QDISC_ENABLE_SIZE 10
> >
> > +#define BLOCKING_FLOW 1
>
> Can you elaborate why this is needed? Is this needed in every shim? Then
> probably it should be defined in a common place (common.h)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/IRATI/stack/pull/713/files#r37167575>.
>

edugrasa replied:
Dont know of my former comment is relevant, sorry if it is not ;)
El 17/08/2015 10:41, "Eduard Grasa Gras" <eduard.grasa@i2cat.net> escribió:

> Blocking or non blocking is a property of flows consumed by apps (via the
> read/write syscalls implented in the KFA), not of flows used by IPCPs.
> El 17/08/2015 10:38, "Sander Vrijders" <notifications@github.com>
> escribió:
>
>> In linux/net/rina/ipcps/shim-eth-vlan.c
>> <https://github.com/IRATI/stack/pull/713#discussion_r37167575>:
>>
>> > @@ -51,6 +51,8 @@
>> >  #define DEFAULT_QDISC_MAX_SIZE 50
>> >  #define DEFAULT_QDISC_ENABLE_SIZE 10
>> >
>> > +#define BLOCKING_FLOW 1
>>
>> Can you elaborate why this is needed? Is this needed in every shim? Then
>> probably it should be defined in a common place (common.h)?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/IRATI/stack/pull/713/files#r37167575>.
>>
>

dstaesse replied:
It's just for readability. if you prefer, I can just put a 
```
bool true /* blocking flow */
```
in the parameter list for the flow allocation. I thing that at some point, there will be be some OPTIONS in the API such as "FLOW_OPT_BLOCKING" and then it will be obvious to replace this in the code.

sandervrijders replied:
No, no, that's ok. My questions are
1) Why is this flow always blocking?
2) Why is this defined every time anew in each shim instead of in a common place?

dstaesse replied:
1) probably we need to add the blocking property to the shim_eth_flow struct then, but usually the app is an IPCP and this is fine (for now)
2) it's not needed in shim-hv, but we can define it anywhere you like.

sandervrijders replied:
1) Yes, but why is it always blocking? I still don't get it.
2) Yes, may be better to have it in a common place imho to avoid code duplication.

dstaesse replied:
2) oh come on... I put it in the .c files so it won't contaminate elsewhere, because if you put it in a header where other files may include it. I'd rather pass "true /* blocking */" than adding this as a definition in a common file. It has to go at some point sooner rather than later and moving the definition outside the file will mean we might miss it.

sandervrijders replied:
Don't see this as a problem. Adding it as an enum in common.h avoids code duplication and makes it re-usable elsewhere if needed.

sandervrijders commented:
General note: Can you please split the one commit into several smaller commits? Now one commit touches everything in the stack. It's better to have one commit for the shims, one for librina etc.

dstaesse commented:
@sander, are you sure? why split in several commits? it doesn't make sense to me because you can't separate this out...

sandervrijders commented:
Of course the granularity of a commit is subjective. But to me, a commit needs to be at a very low level, explaining what files you touched and what you did there. The pull request finally contains all these commits and explains the final solution. Once it is merged in, there will be a hash that we can use to refer to this PR (e.g. Merge xx into pristine-1.3).

But I guess here it's okay to have it in one commit, since it doesn't work otherwise. It's probably because it touches everything that I'm a little weary.

msune commented:
@sandervrijders yes, but commits need to be complete, meaning they must compile and pass the tests. Otherwise we cannot use `git bisect` properly.

sandervrijders commented:
@msune Agreed.

dstaesse commented:
@msune, @sandervrijders : this also shows that it might be premature to split librina, rinad and rina-tools to separate repo's at this point in time. It 's probably better to first consolidate and stabilise the API as much as possible. Else we may have code changes that need transaction-like commits over separate repo's.

msune commented:
We would have the builder for that. THe builder will have all the subrepos via git submodules, which will point to the very exact commit where things work correctly together.

But I tend to agree. At least we need some form of CI to know when things are broken cross-repo.

sandervrijders commented:
Giving it a test.

sandervrijders commented:
```
sander@n:/usr/local/irati/bin$ ./rina-echo-time -s 100 -c 10
4699(1439817812)#librina.logs (DBG): New log level: INFO
4699(1439817812)#librina.nl-manager (INFO): Netlink socket connected to local port 4699 
Flow allocation time = 3.7361 ms
SDU size = 100, seq = 0, RTT = 0.23749 ms
SDU size = 100, seq = 1, RTT = 0.25018 ms
SDU size = 100, seq = 2, RTT = 0.22162 ms
SDU size = 100, seq = 3, RTT = 0.2452 ms
SDU size = 100, seq = 4, RTT = 0.24633 ms
SDU size = 100, seq = 5, RTT = 0.24647 ms
SDU size = 100, seq = 6, RTT = 0.2343 ms
SDU size = 100, seq = 7, RTT = 0.23468 ms
SDU size = 100, seq = 8, RTT = 0.24014 ms
SDU size = 100, seq = 9, RTT = 0.23658 ms
SDUs sent: 10; SDUs received: 10; 0% SDU loss
Minimum RTT: 0.22162 ms; Maximum RTT: 0.25018 ms; Average RTT:0.2393 ms; Standard deviation: 0.0083275 ms
```

Works. ACK.

dstaesse commented:
There's still a bug. Non-blocking I/O doesn't seem to work yet on the accepting side. Gonna investigate further.

dstaesse commented:
Just added a commit cleaning up kernel space. The code compiles, user space compiles as well. I still need to work in user space and add some netlink messages, so some commits are still pending.

sandervrijders commented:
Left some (very minor) comments on the last commits. Apart from that, nice job!

lbergesio commented:
Just want to do a not code-related question, but, why removing my name from the authors' list in shim-tcp-udp and shim-hv. Completly agree I was not the main author there (if it does something I was the last one in the list) but if you do git blame on each file I have 268 and 117 lines respectively and several commits. I was working with them to adapt them to distributed loops and fixing bugs. If this is correct we should revise the criteria for naming authors and recheck the rest of the code. I.e. check Sander ir normal-ipcp.c:

root@irati-vm-leo-1:/home/irati/linux/net/rina# git blame ipcps/normal-ipcp.c | grep "Leonardo" | wc
    451    4625   59530
root@irati-vm-leo-1:/home/irati/linux/net/rina# git blame ipcps/normal-ipcp.c | grep "Sander" | wc
     23     259    3321

sandervrijders commented:
@lbergesio I agree on revising the authorship of the files. We should check what makes most sense. However, I'm not listed in ipcps/normal-ipcp.c as author. I do see myself there as module author but I didn't add that myself:

```
ed5e0e74 linux/net/rina/normal-ipcp.c       (Francesco Salvestrini 2013-10-14 15:29:49 +0200 1323) MODULE_AUTHOR("Sander Vrijders <sander.vrijders@intec.ugent.be>");
```

lbergesio commented:
@sandervrijders not pointing to you or anyone, I just took the first file I got in the same ipcps folder.

dstaesse commented:
@lbergesio : I may have made a mistake, but your contributions are less than 1-2% of the code. According to git blame you have 30 lines of code (2 author lines are removed here) in shim-tcp-udp.c
```
dstaesse@m:~/stack/linux/net/rina/ipcps$ git blame shim-tcp-udp.c | grep Leo | wc -l
30
```
and 28 lines of code in shim-hv (2 author lines removed here as well).
```
dstaesse@m:~/stack/linux/net/rina/ipcps$ git blame shim-hv.c | grep Leo | wc -l
28
```
But the main reason is that I saw that you added your own name, which is not really proper form. Better ask an existing author or the maintainer to add you as an author in such cases.

lbergesio commented:
@dstaesse well, in shim-tcp-udp.c in last pristine-1.3 I can count a bit more:

```root@irati-vm-leo-1:/home/irati/linux/net/rina# git ls-tree -r -z --name-only HEAD -- ipcps/shim-tcp-udp.c | xargs -0 -n1 git blame --line-porcelain HEAD |grep  "^author "|sort|uniq -c|sort -nr
   1127 author Francesco Salvestrini
    939 author Douwe De Bock
    417 author Sander Vrijders
    268 author Leonardo Bergesio
     45 author Eduard Grasa
     11 author Ondrej Lichtner
      4 author Vincenzo Maffione
```

By the time those lines with the author were added we had no maintainers. And for the same reason you said I think it is better to check before removing, more considering you are doing it in a file you are not author of.

vmaffione commented:
It does not make so much sense to count, imho. Authorship is different from contribution. Who designed and first implemented the file is the author (or the authors). Subsequent modifications are contributions, that are tracked by git. But are not authors.

lbergesio commented:
@vmaffione It makes lot of sense to me

dstaesse commented:
@vmaffione: Agree, authorship legally implies parenthood. We probably should add contributors and a maybe a changelog to the headers as well.

In any case, the changes in this PR can be ACK'd or NACK by the maintainers.

vmaffione commented:
That's how it works in an open source community. If I refactor a file, or fix blanks and spaces, or change the API, I'm not the author - I'm a contributor, even if I touch 100% of the file. We may decide to have the list of contributor, I would agree on that.

dstaesse commented:
Fixed the shims according to the discussion, but probably we should have a check on all files, e.g, before finalising 1.3. When 1.3 is final, we could add a CHANGELOG to source files from that point onwards.

dstaesse commented:
PR now works. It contains a nasty hack raping the sduRead and sduWrite syscalls.
However, apps can now decide at will if they want blocking or non-blocking I/O on a flow using
ipcManager->setFlowOptsBlocking(port_id, bool);

```
dstaesse@m:/usr/local/irati/bin$ sudo cp ~/traffic-generator/src/rina-tgen .
dstaesse@m:/usr/local/irati/bin$ ./rina-tgen -l --interval 100
20738(1440697918)#librina.logs (DBG): New log level: INFO
20738(1440697918)#librina.nl-manager (INFO): Netlink socket connected to local port 20738 
20738(1440697955)#traffic-generator (INFO): New flow allocated [port-id = 6]
20738(1440697955)#traffic-generator (INFO): switching to non-blocking I/O
20738(1440697955)#traffic-generator (INFO): Flow IO was 0
20738(1440697955)#traffic-generator (INFO): Flow IO is now 2048
20738(1440697955)#traffic-generator (INFO): Starting test from client traffic.generator.client-1 on port-id 6
20738(1440697955)#traffic-generator (INFO): Duration: 1 s, count: 0 sdus, sdu size: 500 bytes, reporting interval: 100 ms
20738(1440697956)#traffic-generator (INFO): Port    6:        14 SDUs (        7000 bytes) in       100009 us =>  139.9874 p/s,    0.5599 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        20 SDUs (       10000 bytes) in       100010 us =>  199.9800 p/s,    0.7999 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        10 SDUs (        5000 bytes) in       100005 us =>   99.9950 p/s,    0.4000 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        13 SDUs (        6500 bytes) in       100011 us =>  129.9857 p/s,    0.5199 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        30 SDUs (       15000 bytes) in       100012 us =>  299.9640 p/s,    1.1999 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        25 SDUs (       12500 bytes) in       100006 us =>  249.9850 p/s,    0.9999 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        13 SDUs (        6500 bytes) in       100008 us =>  129.9896 p/s,    0.5200 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        21 SDUs (       10500 bytes) in       100006 us =>  209.9874 p/s,    0.8399 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        28 SDUs (       14000 bytes) in       100016 us =>  279.9552 p/s,    1.1198 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6:        26 SDUs (       13000 bytes) in       100007 us =>  259.9818 p/s,    1.0399 Mb/s
20738(1440697956)#traffic-generator (INFO): Port    6: Test complete.
20738(1440697956)#traffic-generator (INFO): switching to blocking I/O
20738(1440697956)#traffic-generator (INFO): Flow IO was 2048
20738(1440697956)#traffic-generator (INFO): Flow IO is now 0
20738(1440697956)#traffic-generator (INFO): Port    6:       200 SDUs,       100000 bytes in      1000 ms, 0.8000 Mb/s
20738(1440697956)#traffic-generator (INFO): Flow torn down remotely [port-id = 6]
```

dstaesse commented:
Did a perf test (no logs), non-blocking read at server, non-blocking write at client.
Non-blocking write needs some debugging in fast path.

For this PR, everything works fine up to at least 400 Mb/s: 
```
4207(1440707345)#traffic-generator (INFO): Port    2:      9998 SDUs (     4999000 bytes) in       100002 us => 99978.0000 p/s,  399.9120 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:      9999 SDUs (     4999500 bytes) in       100001 us => 99989.0000 p/s,  399.9560 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:     10000 SDUs (     5000000 bytes) in       100001 us => 99999.0000 p/s,  399.9960 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:     10003 SDUs (     5001500 bytes) in       100001 us => 100029.0000 p/s,  400.1160 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:     10000 SDUs (     5000000 bytes) in       100001 us => 99999.0000 p/s,  399.9960 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:     10000 SDUs (     5000000 bytes) in       100001 us => 99999.0000 p/s,  399.9960 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:      9999 SDUs (     4999500 bytes) in       100001 us => 99989.0000 p/s,  399.9560 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:     10001 SDUs (     5000500 bytes) in       100001 us => 100009.0000 p/s,  400.0360 Mb/s
4207(1440707345)#traffic-generator (INFO): Port    2:      9996 SDUs (     4998000 bytes) in       100001 us => 99959.0000 p/s,  399.8360 Mb/s
```
trying 500 Mb/s, we run into queuing problems in the fast path:
```
4207(1440707373)#traffic-generator (INFO): Port    2:      8138 SDUs (     4069000 bytes) in       100001 us => 81379.1797 p/s,  325.5168 Mb/s
4207(1440707373)#traffic-generator (INFO): Port    2:      7863 SDUs (     3931500 bytes) in       100001 us => 78629.2109 p/s,  314.5168 Mb/s
4207(1440707373)#traffic-generator (INFO): Port    2:      7855 SDUs (     3927500 bytes) in       100001 us => 78549.2109 p/s,  314.1969 Mb/s
4207(1440707373)#traffic-generator (INFO): Port    2:      8327 SDUs (     4163500 bytes) in       100001 us => 83269.1641 p/s,  333.0767 Mb/s
4207(1440707374)#traffic-generator (INFO): Port    2:      7943 SDUs (     3971500 bytes) in       100002 us => 79428.4141 p/s,  317.7137 Mb/s
4207(1440707374)#traffic-generator (INFO): Port    2:      7889 SDUs (     3944500 bytes) in       100001 us => 78889.2109 p/s,  315.5569 Mb/s
4207(1440707374)#traffic-generator (INFO): Port    2:      7850 SDUs (     3925000 bytes) in       100001 us => 78499.2109 p/s,  313.9969 Mb/s
```

dstaesse commented:
Added a new commit, minimising the API on the allocation side.

dstaesse commented:
Finally found what introduced that bug. Everything seems to be working now.
```
dstaesse@n:/usr/local/irati/bin$ sudo ./rina-echo-time -c 1000 -s 499 -w 2
4749(1441139258)#librina.logs (DBG): New log level: INFO
4749(1441139258)#librina.nl-manager (INFO): Netlink socket connected to local port 4749 
Flow allocation time = 3.8852 ms
SDU size = 499, seq = 0, RTT = 0.51854 ms
SDU size = 499, seq = 1, RTT = 0.52502 ms
...
SDU size = 499, seq = 998, RTT = 0.52147 ms
SDU size = 499, seq = 999, RTT = 0.53519 ms
SDUs sent: 1000; SDUs received: 1000; 0% SDU loss
Minimum RTT: 0.40058 ms; Maximum RTT: 0.65445 ms; Average RTT:0.51542 ms; Standard deviation: 0.031685 ms
```
This PR is now awaiting for ACKS from 
@edugrasa
@lbergesio
@miqueltarzan
@sandervrijders
@vmaffione 
@bernatgaston
@msune
(so, all you guys :-) ).

sandervrijders commented:
Went through the code, looks good, ACK for my parts. Needs to be tested by someone else though.

edugrasa commented:
I don't understand the rationale of why you need a "set/get flow options" operations. If a flow is blocking or not blocking is just one of the characteristics that the application requests for the flow (same as in-order-delivery, minimum characteristics, etc..), so why isn't this just an attribute of the flow specification, communicated by the application at flow allocation time? There's is also an API to retrieve all the flow information already, why do we need get flow options to retrieve if the flow is blocking or not as a separate API?

vmaffione commented:
I completely disagree. The "blocking" property should be a local property of an "I/O handle" (e.g. a file descriptor or a librina flow in our case, etc.), and shouldn't be negotiated with the remote peer (Flow allocation). It is an option that the application sets to decide how to interact with the local kernel, and the value of the option is related to the internal implementation of the application. For example, if the application wants to use an event-loop asynchronous approach, it will use non-blocking I/O (with possibly I/O select, but for now we don't support it). If the applications wants use a simple syncrhonous approach, it will likely use blocking I/O. There are a lots of cases, an application may decide to use a mixture of blocking and non-blocking. But this choice only regards the interaction of the application with the local kernel, the other peer must not be affected by such choice. Communicating applications can have different implementations (e.g. there are different web browsers and different web servers), so may want to manage the I/O in different ways and combinations, but still they have to be able to talk to each other, since they use the same application protocol. This implies that it does not make any sense to "tell" a remote application to use blocking or non-blocking I/O. Each application has to take this decision on this topic - this is not QoS.

edugrasa commented:
Hi vincenzo. From the application point of view, blocking or non-blocking is a property of the flow that needs to be communicated to the IPC Manager, and then to the IPC Process. It doesn't mean that it will be negotiated with the peer, it's just a matter of an information exchange between the application and the IPC Manager (the IPC API is this). Then the IPC Manager (and the IPCP) decide what to do with the implementation. It's not about QoS or not, is about the characteristics of the flow (the flow specification). In our current IPC API, the flow options is the flow information, we don't need to replicate this mechanism. 

When we change to file descriptors, we'll have to redesign the API to make it compliant with file descriptors, but this requires a change of the API as a whole. Meanwhile, let's stick to the current design, which doesn't require introducing new operations and avoids this hack of the read and write syscalls.

msune commented:
Completely agree with Vincenzo here.

The blocking/non-blocking to me is a per-flow thing, which the application can decide depending on how its I/O loop. And this is a application N to application N-1 (in this case application and N-1 IPC in kernel) issue.

edugrasa commented:
Marc, this is not the point, we all agree here (each application decides, App A and B communicating via a flow, A can use blokcing and B non-blocking, no issue with that). I was arguing on how it was implemented, but after a chat with Dimitri, Sander and Vincenzo and checking the code I realised that my idea was not correct, so I'm fine with the current design.

What I think is not acceptable for being merged into 1.3 is the abuse of read/write syscalls to implement the set/get blocking property of a flow. Ok as a quick solution to complete the experiments, but should be implemented properly before being merged into 1.3 in my opinion.

dstaesse commented:
To summarise what's in this PR:
using fd's the workflow is like this
```
fd = open(file);  //opens a file with default I/O access options
fcntl (fd, O_NONBLOCK); //switch to 
```

after this PR, you do this in the IPC api
```
port_id = ipcManager->allocateFlowRequest/Response(); //allocates a flow with default I/O
setFlowOptsBlocking(port_id, blocking); //changes I/O option, now only blocking is implemented
```
the flow_opts goes through a hack in the read/write sdu syscall as I didn't want to add a temporary one just for this. All the other things related to blocking I/O that were in there (in the applicationRegistrationInformation, the FlowSpecification and the allocate_port functions and syscalls) have been removed.
the ultimate goal should be something like:
```
fd = allocateflow(); // allocate a flow
fcntl(fd, O_nonblock);//
```
where I would map the fd and port_id (which are both ints) one-to-one. This looks like a small step from this PR: remove the hack in the read/write syscalls. remove the flowOpts functions from the IPC API and replace with a fcntl() call.

dstaesse commented:
We agreed to scrap the (admittedly, nasty) syscall hack and to implement a dedicated syscall to control flow I/O. It's definitely cleaner, and it will be more extensible as well. 

I added a single syscall flow_io_ctl, where you specify the command FLOW_F_GETFL and FLOW_F_SETFL to get and/or set the options (similar to fcntl).

Added checks to librina build that this syscall is present.

Seems to work:
```
dstaesse@m:/usr/local/irati/bin$ ./rina-tgen -l --interval 100
5232(1441201877)#librina.logs (DBG): New log level: INFO
5232(1441201877)#librina.nl-manager (INFO): Netlink socket connected to local port 5232 
5232(1441201887)#traffic-generator (INFO): New flow allocated [port-id = 2]
5232(1441201887)#traffic-generator (INFO): Starting test from client traffic.generator.client-1 on port-id 2
5232(1441201887)#traffic-generator (INFO): Duration: 5 s, count: 0 sdus, sdu size: 500 bytes, reporting interval: 100 ms
5232(1441201887)#traffic-generator (INFO): Port    2:         1 SDUs (         500 bytes) in       100015 us =>    9.9985 p/s,    0.0400 Mb/s
5232(1441201887)#traffic-generator (INFO): Port    2:         0 SDUs (           0 bytes) in       100020 us =>    0.0000 p/s,    0.0000 Mb/s
5232(1441201888)#traffic-generator (INFO): Port    2:         0 SDUs (           0 bytes) in       100014 us =>    0.0000 p/s,    0.0000 Mb/s
5232(1441201888)#traffic-generator (INFO): Port    2:         2 SDUs (        1000 bytes) in       100041 us =>   19.9918 p/s,    0.0800 Mb/s
5232(1441201888)#traffic-generator (INFO): Port    2:         0 SDUs (           0 bytes) in       100038 us =>    0.00
...
5232(1441201892)#traffic-generator (INFO): Port    2:         0 SDUs (           0 bytes) in       100001 us =>    0.0000 p/s,    0.0000 Mb/s
5232(1441201892)#traffic-generator (INFO): Port    2:         0 SDUs (           0 bytes) in       100001 us =>    0.0000 p/s,    0.0000 Mb/s
5232(1441201892)#traffic-generator (INFO): Port    2:        29 SDUs (       14500 bytes) in       100013 us =>  289.9623 p/s,    1.1598 Mb/s
5232(1441201892)#traffic-generator (INFO): Port    2: Test complete.
5232(1441201892)#traffic-generator (INFO): Port    2:       273 SDUs,       136500 bytes in      5000 ms, 0.2184 Mb/s
5232(1441201892)#traffic-generator (INFO): Flow torn down remotely [port-id = 2]
```
(test done with 40% packet loss)

```
dstaesse@m:~/stack$ sudo cat /var/log/syslog | grep "I/O options"
Sep  2 15:49:27 m kernel: [  314.156085] rina-syscalls(DBG): Got I/O options for port-id = 2: 0
Sep  2 15:49:27 m kernel: [  314.156095] rina-syscalls(DBG): Set I/O options for port-id 2 to 4000
Sep  2 15:51:27 m kernel: [  433.989538] rina-syscalls(DBG): Got I/O options for port-id = 2: 0
Sep  2 15:51:27 m kernel: [  433.989548] rina-syscalls(DBG): Set I/O options for port-id 2 to 4000
Sep  2 15:51:32 m kernel: [  438.993849] rina-syscalls(DBG): Got I/O options for port-id = 2: 4000
Sep  2 15:51:32 m kernel: [  438.993860] rina-syscalls(DBG): Set I/O options for port-id 2 to 0
```

dstaesse commented:
thanks! This PR touches a lot of files all over the place, so we shouldn't linger too long if it's considered as a candidate for merging into pristine-1.3 as to avoid merge conflicts.

sandervrijders commented:
Re-ACKing my parts after latest commit.

edugrasa commented:
Didn't have time to test this morning ;(, but ACKing my parts also.

miqueltarzan commented:
I probably won't have time in the following days to test thoroughly this PR, but I think I can ACK it.

vmaffione commented:
ACK, usual tests work in three-node-scenario
  - rina-echo-time ping and perf
  - rina-tgen

sandervrijders commented:
@bernatgaston, @msune, @lbergesio Can you take a look and ACK as well? I want to merge this in to avoid merge conflicts.

sandervrijders commented:
Just tried some tests as well, works.

msune commented:
@bernatgaston and @lbergesio are on vacation. I ACK only for the files I am maintaining.

edugrasa commented:
ACK on behalf of @bernatgaston and @lbergesio since they are on holidays this week and the pull request should be merged asap since it touches a lot of files
